Python/import.c

```diff
@@ -660,19 +784,32 @@ PyObject *
 PyImport_AddModuleObject(PyObject *name)
 {
     PyObject *modules = PyImport_GetModuleDict();
-    PyObject *m;
+    return _PyImport_AddModuleObject(name, modules);
+}
 
-    if ((m = PyDict_GetItemWithError(modules, name)) != NULL &&
-        PyModule_Check(m)) {
-        return m;
+PyObject *
+_PyImport_AddModuleObject(PyObject *name, PyObject *modules)
+{
+    PyObject *m;
+    if (PyDict_Check(modules)) {
+        m = PyDict_GetItemWithError(modules, name);
+    } else {
+        m = PyObject_GetItem(modules, name);
+        // For backward-comaptibility we copy the behavior
+        // of PyDict_GetItemWithError().
+        if (PyErr_Occurred() && !PyMapping_HasKey(modules, name))
+            PyErr_Clear();
     }
     if (PyErr_Occurred()) {
```

Comment:
Hum, maybe merge this test with the previous test (don't test PyErr_Occurred twice).

Reply:
I've changed this to use PyErr_ExceptionMatches(), so I'd rather leave the rest alone.